core/src/main/scala/scalaz/IList.scala

```diff
@@ -434,6 +434,18 @@ sealed abstract class IList[A] extends Product with Serializable {
   def toZipper: Option[Zipper[A]] =
     sToZipper(toStream)
 
+  // optimised specialisation of Traverse.traverse for disjunction, avoiding
+  // allocations for each element, approximately 20% faster than the default
+  // implementation.
+  def traverseDisjunction[E, B](f: A => E \/ B): E \/ IList[B] = \/-(
+    map { a =>
+      f(a) match {
+        case -\/(err) => return -\/(err)
```

Comment:
write a tailrec function instead of using exceptions?

Reply:
Also `map` runs in the opposite direction:
```scala
   def map[B](f: A => B): IList[B] = reverse.reverseMap(f)
```
It starts from the end, so your `traverse` is not the same as the usual one.
^^ a perfect example why side-effects are evil ;)

Reply:
```scala
def traverse[E, A, B](l: List[A], f: A => E Either B, r: List[B]): E Either List[B] =
  l match {
    case a :: as =>
      f(a) match {
        case Left(e) => Left(e)
        case Right(b) => traverse(as, f, b :: r)
      }
    case Nil => Right(r.reverse)
  }

println(
  traverse(
    List(1, 2, 3, 4), 
    (a : Int) => if (a % 2 == 0) Left(a) else Right(a), 
    Nil))
```